bin/propolis-server/src/lib/server.rs

```diff
@@ -318,36 +342,127 @@ fn instance_spec_from_request(
     Ok(VersionedInstanceSpec::V0(spec_builder.finish()))
 }
 
-/// Attempts to register an Oximeter server reporting metrics from a new
-/// instance, returning the producer registry from that server on success.
-async fn register_oximeter(
-    server_context: &DropshotEndpointContext,
-    cfg: &MetricsEndpointConfig,
-    vm_id: uuid::Uuid,
+/// Register an Oximeter server reporting metrics from a new instance.
+///
+/// This spawns a tokio task which will indefinitely attempt to register with
+/// Nexus as an Oximeter metric producer. Once the registration succeeds, it
+/// will also arrange for the statistics produced for the managed instance to be
+/// collected.
+async fn register_oximeter_in_background(
+    services: Arc<ServiceProviders>,
+    cfg: MetricsEndpointConfig,
+    registry: ProducerRegistry,
+    virtual_machine: VirtualMachine,
     log: Logger,
-) -> anyhow::Result<ProducerRegistry> {
-    assert!(server_context.services.oximeter_stats.lock().await.is_none());
-    assert!(server_context
-        .services
-        .oximeter_server_task
-        .lock()
+) {
+    assert!(services.oximeter_state.stats.lock().await.is_none());
+    assert!(services.oximeter_state.server_task.lock().await.is_none());
+    assert!(services.oximeter_state.registration_task.lock().await.is_none());
+
+    // Spin indefinitely awaiting to register with Nexus as a metric producer.
+    //
+    // This either returns a server, or None, indicating we hit a non-retryable
+    // error. We'll bail early in that case, since registering our own metrics
+    // is pointless.
+    let services_ = services.clone();
+    let registration_task = tokio::task::spawn(async move {
+        // Register with Nexus, backing off reasonably quickly if we hit a
+        // retryable error.
+        let register_with_nexus = || async {
+            crate::stats::start_oximeter_server(
+                virtual_machine.instance_id,
+                &cfg,
+                &log,
+                &registry,
+            )
+            .await
+            .map_err(|e| match e {
+                oximeter_producer::Error::RegistrationError {
+                    retryable,
+                    msg,
+                } if retryable => {
+                    omicron_common::backoff::BackoffError::transient(msg)
+                }
+                _ => omicron_common::backoff::BackoffError::permanent(
+                    e.to_string(),
+                ),
+            })
+        };
```

Comment:
Should `OximeterState` also hold some sort of notification channel to allow `register_with_nexus` or the backoff loop to bail out promptly if the server is being shutdown?

Reply:
Sure, I can do that, and signal it where we shutdown all the provider tasks.

Reply:
Done in e295e14